packages/@cdktf/provider-generator/lib/get/generator/models/attribute-model.ts

```diff
@@ -158,20 +158,30 @@ export class AttributeModel {
     );
   }
 
-  public get setterType(): SetterType {
-    return this.isStored
-      ? this.getterType._type === "stored_class"
-        ? {
-            _type: "put",
-            type: this.type.storedName,
-          }
-        : {
-            _type: "set",
-            type: `${this.type.storedName}${
-              this.isProvider ? "| undefined" : ""
-            }`,
-          }
-      : { _type: "none" };
+  public setterType(context?: "resource" | "struct"): SetterType {
+    if (!this.isStored) {
+      return { _type: "none" };
+    }
+
+    if (this.getterType._type === "stored_class") {
+      return {
+        _type: "put",
+        type: this.type.storedName,
+      };
+    }
+
+    // TODO: this is a hack to make optional attributes required on resources but not on structs
+    if (context === "resource") {
+      return {
+        _type: "set",
+        type: `${this.type.name}${this.isProvider ? "| undefined" : ""}`,
+      };
+    }
+
+    return {
+      _type: "set",
+      type: `${this.type.storedName}${this.isProvider ? "| undefined" : ""}`,
```

Comment:
I'm not sure I follow why structs should be different.

Reply:
Because we currently support setters in structs, leading to different types for getters and setters on these. There is a bigger solution that should replace this temporary "make it the old thing in this case" extra case

Reply:
I removed this, not sure why I had it in the first place. If something arises I'll add a test to catch it 👍